src/main/java/org/elasticsearch/index/query/QueryBuilder.java

```diff
@@ -37,6 +39,57 @@
  */
 public abstract class QueryBuilder<QB extends QueryBuilder> extends ToXContentToBytes implements Writeable<QB> {
 
+    public static enum Operator {
+        OR, AND;
+
+        public BooleanClause.Occur toMustOrShouldClause() {
+            switch (this) {
+                case OR:
+                    return BooleanClause.Occur.SHOULD;
+                case AND:
+                    return BooleanClause.Occur.MUST;
+                default:
+                    throw Operator.newOperatorException(this.toString());
+            }
+        }
+
+        public static Operator readFrom(StreamInput in) throws IOException {
+            switch (in.readVInt()) {
+                case 0:
```

Comment:
tiny little comment: we could save the ordinal as part of the Operator enum, then we would have the value in a single place. We could loop over the enum values and compare the ordinal with the readVInt result to return the proper Operator. Below, in writeTo, we could simply do out.writeVInt(operator.id());, no need for the switch.